File: src/main/java/de/schramm/royalbash/tdd/core/creature/CreatureUtil.java
package de.schramm.royalbash.tdd.core.creature;

import de.schramm.royalbash.tdd.core.Context;
import de.schramm.royalbash.tdd.core.Creature;
import de.schramm.royalbash.tdd.core.Game;
import lombok.val;

class CreatureUtil {

    static Game spawnCreature(Creature creature, Context context) {

        val game = context.getGame();
        val owner = context.getOwner();
        val targetSpot = context.getTargetSpot();

        val updatedSpot = targetSpot.place(creature);

        return game.getPlayer(owner)
                .map(player -> player.updateSpot(targetSpot, updatedSpot))
                .map(player -> game.updatePlayer(owner, player))
                .orElse(game);
    }
}

File: backend/src/main/kotlin/de/schramm/royalbash/domain/card/creature/CreatureUtil.kt
package de.schramm.royalbash.domain.card.creature

import de.schramm.royalbash.domain.*

internal object CreatureUtil {

    fun spawnCreature(creature: Creature, context: Context): Game {

        val game = context.game
        val owner = context.owner

        return context.targetSpot
                ?.let { targetSpot -> game.findPlayer(owner)
                        ?.let { it.updateSpot(targetSpot to targetSpot.place(creature)) }
                        ?.let { game.updatePlayer(owner to it) }
                        ?: game }
                ?: game
    }
}
